Expand and revamp the deployments docs (#2087)
mendersoftware/mender-docs: 02.Overview/05.Deployment/docs.md

    category: docs
---

A Deployment ensures delivery of a Release to one or more devices. A Release can contain one or more Mender Artifacts.
The Artifact is assigned to a device based on the software and hardware compatibility the device provided.
In other words, a Deployment is the server side model which represent a device update.

The characteristics of a Deployment are:
* Only accepted devices can be part of a Deployment

> * A device won't take part in the same Deployment twice

Reviewer: I think that’s wrong wording. It also won‘t take part in the Deployment twice, but it can receive the Artifact twice. And even that only applies to the „is already installed“ case. My understanding is that the „force“ option should not be mentioned in this context.

Author: Sorry, not getting this.
Can you try rephrasing?

### Deployment to static groups

A Deployment to a static group contains a fixed list of devices and finishes once all the devices in the Deployment have finished.
Mender Server translates the group into a fixed list of Device IDs at Deployment creation time.

A Deployment to static groups will:

* Finish as soon as all devices have either reported success or failure
* Apply to a fixed number of devices during the lifetime of the Deployment
    * Adding or removing devices to the static group after the Deployment creation has no effect on the ongoing Deployment.

### Deployment to dynamic groups

In the Enterprise plan, Mender Server supports the creation of Deployments using a dynamic filter to assign a Deployment to an open, variable group of devices matching a filter criteria.

A Deployment to a dynamic group behaves differently than to a static group.

A Deployment to dynamic groups will:

* Never finish unless the user manually stops it or it was explicitly capped to a limited number of devices
* The number of devices taking part in the Deployment can change during the lifetime of the Deployment

> #### Special case Deployment to dynamic groups

Reviewer: Why do we need a heading here, if there is no second section on the same level? Plus, the title does not sit right with me. Maybe better „Example: Deployment order on dynamic groups“

Author: Like the new title.
Changing.

### Phased rollouts and dynamic groups

The calculation for the number of devices in a phases uses the total number of devices in the dynamic group at the moment of Deployment creation.
The nature of Deployments to a dynamic group is that the total number of devices receiving the update can change during the lifetime of the Deployment.

If the number of devices in a dynamic group changes during a phased rollout the original rollout calculation won't be recalculated.
Depending of the nature of the change the following will happen:
* the number of devices in the dynamic group increases - last phase expands to contain all remaining device.
* the number of devices in the dynamic group decreases - the entire group (potentially) gets updated before all the phases end

!! The exact time a specific device receives a deployed Artifact is not determined by the server, but the concrete moment the device contacts the server following its polling cycle.
!! This means it is possible that devices which joined the group after the Deployment started, get updated as part of the first phase.

> 

Reviewer: And then this could be another „### Example: phases of Deployment to dynamic groups“

Author: Yes! Fixing

    category: docs
---

A Deployment ensures delivery of a Release to one or more devices. A Release can contain one or more Mender Artifacts.
The Artifact is assigned to a device based on the software and hardware compatibility the device provided.

> In other words, a Deployment is the server side model which represent a device update.

Reviewer 2: A Deployment ensures the delivery of a Release to one or more devices. A Release can contain one or more Mender Artifacts.
The Artifact is assigned to a device based on the software and hardware compatibility the device provides.
In other words, a Deployment is the server-side model which represents a device update.

Author: Applying

The characteristics of a Deployment are:
* Only accepted devices can be part of a Deployment
* A device won't take part in the same Deployment twice
    * This can be overridden with the "Force update" option
* Deployments get applied to the device in the chronological order they are created

### Deployment to static groups

A Deployment to a static group contains a fixed list of devices and finishes once all the devices in the Deployment have finished.

> Mender Server translates the group into a fixed list of Device IDs at Deployment creation time.

Reviewer 2: ```suggestion
The Mender Server translates the group into a fixed list of Device IDs at Deployment creation time.
```

Author: Applying manually as commiting from github messes up the author and I need to manually fix ig

A Deployment to static groups will:

> * Finish as soon as all devices have either reported success or failure
> * Apply to a fixed number of devices during the lifetime of the Deployment
>     * Adding or removing devices to the static group after the Deployment creation has no effect on the ongoing Deployment.

Reviewer 2: ```suggestion
A Deployment to static groups will:
* Finish as soon as all devices have either reported success or failure
* Apply to a specified number of devices during the lifetime of the Deployment
    * Adding or removing devices to the static group after the Deployment creation does not affect the ongoing Deployment.
```

### Deployment to dynamic groups

> In the Enterprise plan, Mender Server supports the creation of Deployments using a dynamic filter to assign a Deployment to an open, variable group of devices matching a filter criteria.

Reviewer 2: ```suggestion
In the Enterprise plan, Mender Server supports the creation of Deployments using a dynamic filter to assign a Deployment to an open, variable group of devices matching a filter criterion.
```

> A Deployment to a dynamic group behaves differently than to a static group.
> 
> A Deployment to dynamic groups will:

Reviewer 2: ```suggestion
A Deployment to a dynamic group behaves differently than to a static group. In particular, it will:
```

It's applied to the device, the device gets updated to v2. This results in a group change so the device is now part of `group-v2`.

Deployment `(2.)` was created after Deployment `(1.)` so it can be applied to the device next and the device updates to v3.

Let's examine the same case:

```
group-v1 - devices with v1      # an example device starts with the version v1
group-v2 - devices with v2
group-v3 - devices with v3
```

But in this case the Deployment creation order is reversed:

    (1.) Deploy v3 to group-v2
    (2.) Deploy v2 to group-v1

Deployment (1.) is created first, but the device isn't part of group-v2.
Instead Deployment (2.) takes place and updates the device to v2.

The device is now part of group-v2 and and there is an ongoing Deployment (1.) to group-v2.
However Deployment (1.) **will not start** for the device to not break the guaranteed ordering rule.

From the perspective of the device Deployment (1.) has to happen before Deployment (2.),
Since Deployment (2.) already took place there is no possible way for (1.) to take place without breaking the guarantee.

### Phased rollouts and dynamic groups

> The calculation for the number of devices in a phases uses the total number of devices in the dynamic group at the moment of Deployment creation.

Reviewer 2: ```suggestion
The calculation for the number of devices in a phase uses the total number of devices in the dynamic group at the moment of Deployment creation.
```

> The nature of Deployments to a dynamic group is that the total number of devices receiving the update can change during the lifetime of the Deployment.

Reviewer 2: ```suggestion
The nature of Deployments to dynamic groups is that the total number of devices receiving the update can change during the lifetime of the Deployment.
```
